pkg/controller/oam/applicationconfiguration/render.go

```diff
@@ -40,15 +40,20 @@ const (
 
 // Render error format strings.
 const (
-	errFmtGetComponent     = "cannot get component %q"
-	errFmtResolveParams    = "cannot resolve parameter values for component %q"
-	errFmtRenderWorkload   = "cannot render workload for component %q"
-	errFmtRenderTrait      = "cannot render trait for component %q"
-	errFmtSetParam         = "cannot set parameter %q"
-	errFmtUnsupportedParam = "unsupported parameter %q"
-	errFmtRequiredParam    = "required parameter %q not specified"
+	errFmtGetComponent          = "cannot get component %q"
+	errFmtGetWorkloadDefinition = "cannot get workloadDefinition %q"
+	errFmtResolveParams         = "cannot resolve parameter values for component %q"
+	errFmtRenderWorkload        = "cannot render workload for component %q"
+	errFmtRenderTrait           = "cannot render trait for component %q"
+	errFmtSetParam              = "cannot set parameter %q"
+	errFmtUnsupportedParam      = "unsupported parameter %q"
+	errFmtRequiredParam         = "required parameter %q not specified"
 )
 
+// check components implements ComponentRenderer
+// there is no other way in Golang to explicitly force implement an interface
+var _ ComponentRenderer = &components{}
```

Comment:
nitpick: I usually prefer to do interface checks in `test` file

Reply:
One of my major complaints about Golang is there is no way to enforce "isA" relation and this is the hack version of it. I think this improves the code readability dramatically.